aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,607 @@
+  max_unpooling3d_forward_out_cpu(
+      output, self, indices, output_size, stride, padding);
+  return output;
+}
+
+template <typename scalar_t>
+static void max_unpooling2d_backward_out_cpu_frame(
+    scalar_t* gradInput_p,
+    scalar_t* gradOutput_p,
+    int64_t* ind_p,
+    int64_t nslices,
+    int64_t iwidth,
+    int64_t iheight,
+    int64_t owidth,
+    int64_t oheight) {
+  bool has_error = false;
+  int64_t error_index = 0;
+  int k;
+#pragma omp parallel for private(k)
+  for (k = 0; k < nslices; k++) {
+    scalar_t* gradInput_p_k = gradInput_p + k * iwidth * iheight;
+    scalar_t* gradOutput_p_k = gradOutput_p + k * owidth * oheight;
+    int64_t* ind_p_k = ind_p + k * iwidth * iheight;
+
+    int64_t i, j;
+    int64_t maxp;
+
+    for (i = 0; i < iheight; i++) {
+      for (j = 0; j < iwidth; j++) {
+        maxp = ind_p_k[i * iwidth + j]; /* retrieve position of max */
+        if (maxp < 0 || maxp >= owidth * oheight) {
+#pragma omp critical
+          {
+            has_error = true;
+            error_index = maxp;
+          }
+        }
+        gradInput_p_k[i * iwidth + j] =
+            gradOutput_p_k[maxp]; /* update gradient */
+      }
+    }
+  }
+  if (has_error) {
+    AT_ERROR(
+        "invalid max index ",
+        error_index,
+        ", owidth= ",
+        owidth,
+        ", oheight= ",
+        oheight);
+  }
+}
+
+Tensor& max_unpooling2d_backward_out_cpu(
+    Tensor& grad_input,
+    const Tensor& grad_output_,
+    const Tensor& self,
+    const Tensor& indices_,
+    IntList output_size) {
+  int64_t oheight = output_size[0];
+  int64_t owidth = output_size[1];
+  int dimw = 2;
+  int dimh = 1;
+  int nbatch = 1;
+  int nslices;
+  int iheight;
+  int iwidth;
+  AT_CHECK(
+      indices_.scalar_type() == at::ScalarType::Long,
+      "elements in indices should be type int64");
+  AT_CHECK(
+      self.sizes() == indices_.sizes(), "Input shape must match indices shape");
+
+  AT_CHECK(output_size.size() == 2, "Output size must be 2");
+
+  /* get contiguous gradOutput and indices */
+  auto grad_output = grad_output_.contiguous();
+  auto indices = indices_.contiguous();
+
+  /* resize */
+  grad_input.resize_as_(self);
+  grad_input.zero_();
+
+  if (self.ndimension() == 4) {
+    nbatch = self.size(0);
+    dimw++;
+    dimh++;
+  }
+
+  /* sizes */
+  nslices = self.size(dimh - 1);
+  iheight = self.size(dimh);
+  iwidth = self.size(dimw);
+
+  if (owidth != grad_output.size(dimw) || oheight != grad_output.size(dimh)) {
+    AT_ERROR(
+        "Inconsistent gradOutput size. output height = ",
+        oheight,
+        ", output width = ",
+        owidth,
+        ", gradOutput: ",
+        grad_output.size(dimh),
+        "x",
+        grad_output.size(dimw));
+  }
+
+  int p;
+  for (p = 0; p < nbatch; p++) {
+    auto inputOffset = p * nslices * iheight * iwidth;
+    auto outputOffset = p * nslices * oheight * owidth;
+    AT_DISPATCH_FLOATING_TYPES(
```

Comment:
You're doing a dynamic dispatch for every batch in the tensor, when you could just do the dynamic dispatch once.